### PR TITLE
Feat/distinct field

### PR DESCRIPTION
**Pull request by anik-ghosh-au7**
Support and examples added for `distinctField `& `distinctFieldConfig` props in flutter-searchbox - which returns distinct results based on a specified field.

Notion Link - https://www.notion.so/appbase/distinctField-prop-f14649fd05c544a993cbee11e1846c8c

Docs PR link - https://github.com/appbaseio/Docs/pull/186

Demo - 
https://www.loom.com/share/558c4f8619ba43ccaa35b4f3c50450d4

### CODE REVIEW

File: searchbase/lib/src/searchcontroller.dart [hunk 1]
```diff
@@ -337,6 +337,12 @@ class SearchController extends Base {
   /// A list of recent searches as suggestions.
   List<Suggestion>? recentSearches;
 
+  /// A string defining a field's name, used to return the distinct value documents.
```
Comment (bietkul) [inline searchbase/lib/src/searchcontroller.dart]: @anik-ghosh-au7 Define docs with example.

File: searchbase/lib/src/searchcontroller.dart [hunk 2]
```diff
@@ -337,6 +337,12 @@ class SearchController extends Base {
...
+  String? distinctField;
+
+  /// This property allows specifying additional options to the distinctField property.
```
Comment (bietkul) [inline searchbase/lib/src/searchcontroller.dart]: @anik-ghosh-au7 Define docs with example.

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 3]
```diff
@@ -1735,6 +1781,39 @@ class SearchBox<S, ViewModel> extends SearchDelegate<String?> {
   /// Data must be in form of Elasticsearch response.
   final List<Map>? results;
 
+  /// A string defining a field's name, used to return the distinct value documents.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]: @anik-ghosh-au7 Why are you not using the content from docs.appbase.io?
Please use the exact content with links and examples.
alt img

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 4]
```diff
@@ -1735,6 +1781,39 @@ class SearchBox<S, ViewModel> extends SearchDelegate<String?> {
...
+  /// For example,
+  /// ```dart
+  /// SearchBox(
+  ///   ...
+  ///   distinctField: 'authors.keyword',
+  /// )
+  /// ```
+  final String? distinctField;
+
+  /// This property allows specifying additional options to the distinctField property.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]: @anik-ghosh-au7 Fix content

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 5]
```diff
@@ -1196,6 +1204,39 @@ class SearchWidgetConnector<S, ViewModel> extends StatelessWidget {
   /// ```
   final TransformResponse? transformResponse;
 
+  /// A string defining a field's name, used to return the distinct value documents.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]: Fix content

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 6]
```diff
@@ -1196,6 +1204,39 @@ class SearchWidgetConnector<S, ViewModel> extends StatelessWidget {
...
+  /// For example,
+  /// ```dart
+  /// SearchBox(
+  ///   ...
+  ///   distinctField: 'authors.keyword',
+  /// )
+  /// ```
+  final String? distinctField;
+
+  /// This property allows specifying additional options to the distinctField property.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]: Fix content

File: flutter_searchbox/example_with_distinctField/lib/main.dart [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                    // SearchBox widget from flutter searchbox
+                    delegate: SearchBox(
+                      // A unique identifier that can be used by other widgets to reactively update data
+                      id: 'search-widget',
+                      size: 5,
+                      dataField: [
+                        {'field': 'original_title', 'weight': 1},
+                        {'field': 'original_title.search', 'weight': 3}
+                      ],
+                      distinctField: 'authors.keyword',
```
Comment (bietkul) [inline flutter_searchbox/example_with_distinctField/lib/main.dart]: @anik-ghosh-au7 Remove it from here

File: flutter_searchbox/example_with_distinctField/lib/main.dart [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                      distinctFieldConfig: {
```
Comment (bietkul) [inline flutter_searchbox/example_with_distinctField/lib/main.dart]: @anik-ghosh-au7 Remove it from here

File: flutter_searchbox/example_with_distinctField/lib/results.dart [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+
+class ResultsWidget extends StatelessWidget {
+  final SearchController searchController;
+  Map<String, dynamic> otherBooks = {};
+
+  ResultsWidget(this.searchController);
+  @override
+  Widget build(BuildContext context) {
+    var books;
+    searchController.results?.rawData?['hits']['hits'].forEach((hit) => {
```
Comment (bietkul) [inline flutter_searchbox/example_with_distinctField/lib/results.dart:35]: @anik-ghosh-au7 What are you doing here? Can you write some comments?

**Review by bietkul — CHANGES_REQUESTED** [review]
@anik-ghosh-au7 Fix the request changes.
How can you reach the end of the results? It seems like an issue with the `No more results` message.
alt img

**Comment by siddharthlatest** [comment]
@anik-ghosh-au7 I would also add that `.idea/` dir files should be ignored (see .gitignore) and not be part of the PR.

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 10]
```diff
@@ -1196,6 +1204,35 @@ class SearchWidgetConnector<S, ViewModel> extends StatelessWidget {
   /// ```
   final TransformResponse? transformResponse;
 
+  /// String [optional] This prop returns only the distinct value documents for the specified field.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]:
```suggestion
  /// This prop returns only the distinct value documents for the specified field.
```

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 11]
```diff
@@ -1196,6 +1204,35 @@ class SearchWidgetConnector<S, ViewModel> extends StatelessWidget {
...
+  /// It is equivalent to the DISTINCT clause in SQL. It internally uses the collapse feature of Elasticsearch.
+  /// You can read more about it over here - https://www.elastic.co/guide/en/elasticsearch/reference/current/collapse-search-results.html
+  final String? distinctField;
+
+  /// Object [optional] This prop allows specifying additional options to the distinctField prop.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]:
```suggestion
  /// This prop allows specifying additional options to the distinctField prop.
```

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 12]
```diff
@@ -1735,6 +1777,35 @@ class SearchBox<S, ViewModel> extends SearchDelegate<String?> {
   /// Data must be in form of Elasticsearch response.
   final List<Map>? results;
 
+  /// String [optional] This prop returns only the distinct value documents for the specified field.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]:
```suggestion
  /// This prop returns only the distinct value documents for the specified field.
```

File: flutter_searchbox/lib/flutter_searchbox.dart [hunk 13]
```diff
@@ -1735,6 +1777,35 @@ class SearchBox<S, ViewModel> extends SearchDelegate<String?> {
...
+  /// It is equivalent to the DISTINCT clause in SQL. It internally uses the collapse feature of Elasticsearch.
+  /// You can read more about it over here - https://www.elastic.co/guide/en/elasticsearch/reference/current/collapse-search-results.html
+  final String? distinctField;
+
+  /// Object [optional] This prop allows specifying additional options to the distinctField prop.
```
Comment (bietkul) [inline flutter_searchbox/lib/flutter_searchbox.dart]:
```suggestion
  /// This prop allows specifying additional options to the distinctField prop.
```

File: searchbase/lib/src/searchcontroller.dart [hunk 14]
```diff
@@ -337,6 +337,35 @@ class SearchController extends Base {
   /// A list of recent searches as suggestions.
   List<Suggestion>? recentSearches;
 
+  /// String [optional] This prop returns only the distinct value documents for the specified field.
```
Comment (bietkul) [inline searchbase/lib/src/searchcontroller.dart]:
```suggestion
  /// This prop returns only the distinct value documents for the specified field.
```

File: searchbase/lib/src/searchcontroller.dart [hunk 15]
```diff
@@ -337,6 +337,35 @@ class SearchController extends Base {
...
+  /// It is equivalent to the DISTINCT clause in SQL. It internally uses the collapse feature of Elasticsearch.
+  /// You can read more about it over here - https://www.elastic.co/guide/en/elasticsearch/reference/current/collapse-search-results.html
+  final String? distinctField;
+
+  /// Object [optional] This prop allows specifying additional options to the distinctField prop.
```
Comment (bietkul) [inline searchbase/lib/src/searchcontroller.dart]:
```suggestion
  /// This prop allows specifying additional options to the distinctField prop.
```

**Review by bietkul — APPROVED** [review]

**Pull request merged**